WalletWasabi/Crypto/ZeroKnowledge/NonInteractive/IKnowledge.cs

```diff
@@ -0,0 +1,21 @@
+using WalletWasabi.Crypto.Randomness;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.NonInteractive
+{
+	// IKnowledge splits proving into 3 phases
+	// The second phase is to generate and commit to all the nonces
+	public delegate RespondToChallenge ProverCommitToNonces(WasabiRandom random);
+
+	// The third phase is to generate challenges and respond to them
+	public delegate Proof RespondToChallenge();
+
+	public interface IKnowledge
```

Comment:
```suggestion
	/// <summary>
	/// <see cref="IKnowledge"/> splits proving into 3 phases:
	/// <list type="number">
	/// <item>The first phase is to commit to all the statements using <see cref="CommitToStatements(Transcript)"/>,
	/// so that synthetic nonce generation for every sub-proof depends on the statement as a whole.</item>
	/// <item>The second phase is to generate and commit to all the nonces using <see cref="ProverCommitToNonces"/>.</item>
	/// <item>The third phase is to generate challenges using <see cref="RespondToChallenge"/> and respond to them.</item>
	/// </list>
	/// </summary>
	public interface IKnowledge
```

Reply:
I will apply this to `IStatement` as well